File: rust/src/runtime/mcp_proxy_plane.rs
use std::collections::{HashMap, HashSet};
use std::pin::Pin;
use std::sync::Arc;

use async_stream::try_stream;
use deepstrike_core::types::message::{ToolCall, ToolSchema};
use futures::stream::Stream;
use futures::StreamExt;
use serde_json::Value;
use tokio::io::{AsyncBufReadExt, AsyncWriteExt, BufReader};
use tokio::sync::Mutex;

use crate::run_event::RunEvent;
use crate::runtime::credential_vault::CredentialVault;
use crate::runtime::execution_plane::{ExecutionPlane, LocalExecutionPlane, RunContext};
use crate::tools::RegisteredTool;
use crate::{Error, Result};

// ── Server config ─────────────────────────────────────────────────────────────

pub struct McpServerConfig {
    /// Executable to run (e.g. "npx", "python3", "/usr/local/bin/my-mcp-server").
    pub command: String,
    pub args: Vec<String>,
    /// Vault keys injected as env vars into the subprocess. Never exposed to the model.
    pub credential_keys: Vec<String>,
    /// Additional static env vars forwarded to the subprocess.
    pub env: HashMap<String, String>,
}

// ── Internal MCP connection ───────────────────────────────────────────────────

struct Inner {
    stdin: tokio::process::ChildStdin,
    reader: BufReader<tokio::process::ChildStdout>,
    next_id: u64,
}

struct McpConnection {
    server_name: String,
    inner: Mutex<Inner>,
    child: Mutex<Option<tokio::process::Child>>,
    schemas: Vec<ToolSchema>,
}

async fn do_request(
    inner: &mut Inner,
    server_name: &str,
    method: &str,
    params: Option<Value>,
) -> Result<Value> {
    let id = inner.next_id;
    inner.next_id += 1;

    let mut msg = serde_json::json!({ "jsonrpc": "2.0", "method": method, "id": id });
    if let Some(p) = params {
        msg["params"] = p;
    }
    let line = serde_json::to_string(&msg).map_err(|e| Error::Other(e.to_string()))? + "\n";

    inner.stdin.write_all(line.as_bytes()).await.map_err(|e| Error::Other(e.to_string()))?;
    inner.stdin.flush().await.map_err(|e| Error::Other(e.to_string()))?;

    loop {
        let mut buf = String::new();
        let n = inner.reader.read_line(&mut buf).await.map_err(|e| Error::Other(e.to_string()))?;
        if n == 0 {
            return Err(Error::Other(format!("MCP server '{server_name}' disconnected")));
        }
        let trimmed = buf.trim();
        if trimmed.is_empty() {
            continue;
        }
        let Ok(resp) = serde_json::from_str::<Value>(trimmed) else { continue };
        // Skip notifications (no numeric id)
        let Some(resp_id) = resp.get("id").and_then(Value::as_u64) else { continue };
        if resp_id != id {
            continue;
        }
        if let Some(err) = resp.get("error") {
            return Err(Error::Other(format!("MCP({server_name}) error: {err}")));
        }
        return Ok(resp.get("result").cloned().unwrap_or(Value::Null));
    }
}

async fn do_notify(inner: &mut Inner, method: &str) {
    let msg = serde_json::json!({ "jsonrpc": "2.0", "method": method });
    if let Ok(s) = serde_json::to_string(&msg) {
        inner.stdin.write_all((s + "\n").as_bytes()).await.ok();
        inner.stdin.flush().await.ok();
    }
}

impl McpConnection {
    async fn start(
        server_name: &str,
        config: &McpServerConfig,
        vault: &dyn CredentialVault,
    ) -> Result<Arc<Self>> {
        let mut env: HashMap<String, String> = std::env::vars().collect();
        for (k, v) in &config.env {
            env.insert(k.clone(), v.clone());
        }
        for key in &config.credential_keys {
            if let Some(val) = vault.get(key).await {
                env.insert(key.clone(), val);
            }
        }

        let mut child = tokio::process::Command::new(&config.command)
            .args(&config.args)
            .envs(&env)
            .stdin(std::process::Stdio::piped())
            .stdout(std::process::Stdio::piped())
            .stderr(std::process::Stdio::inherit())
            .spawn()
            .map_err(|e| Error::Other(format!("failed to spawn MCP server '{server_name}': {e}")))?;

        let stdin = child
            .stdin
            .take()
            .ok_or_else(|| Error::Other(format!("MCP server '{server_name}': no stdin handle")))?;
        let stdout = child
            .stdout
            .take()
            .ok_or_else(|| Error::Other(format!("MCP server '{server_name}': no stdout handle")))?;

        let mut inner = Inner { stdin, reader: BufReader::new(stdout), next_id: 1 };

        // MCP handshake
        do_request(
            &mut inner,
            server_name,
            "initialize",
            Some(serde_json::json!({
                "protocolVersion": "2024-11-05",
                "capabilities": { "tools": {} },
                "clientInfo": { "name": "deepstrike", "version": "0.1.0" },
            })),
        )
        .await?;
        do_notify(&mut inner, "notifications/initialized").await;

        // Discover tool schemas
        let list = do_request(&mut inner, server_name, "tools/list", None).await?;
        let raw_tools = list.get("tools").and_then(Value::as_array).cloned().unwrap_or_default();

        let mut schemas = Vec::new();
        for t in &raw_tools {
            let name = t.get("name").and_then(Value::as_str).unwrap_or("");
            if name.is_empty() {
                continue;
            }
            let description = t.get("description").and_then(Value::as_str).unwrap_or(name);
            let parameters = t
                .get("inputSchema")
                .cloned()
                .unwrap_or_else(|| serde_json::json!({ "type": "object", "properties": {} }));
            schemas.push(ToolSchema {
                name: name.into(),
                description: description.to_string(),
                parameters,
            });
        }

        Ok(Arc::new(McpConnection {
            server_name: server_name.to_string(),
            inner: Mutex::new(inner),
            child: Mutex::new(Some(child)),
            schemas,
        }))
    }

    async fn execute(&self, call: &ToolCall) -> (String, bool) {
        let params = serde_json::json!({
            "name": call.name.as_str(),
            "arguments": call.arguments,
        });
        let mut inner = self.inner.lock().await;
        match do_request(&mut inner, &self.server_name, "tools/call", Some(params)).await {
            Ok(result) => {
                let text = result
                    .get("content")
                    .and_then(Value::as_array)
                    .map(|arr| {
                        arr.iter()
                            .filter(|c| c.get("type").and_then(Value::as_str) == Some("text"))
                            .filter_map(|c| c.get("text").and_then(Value::as_str))
                            .collect::<Vec<_>>()
                            .join("\n")
                    })
                    .unwrap_or_default();
                let is_error =
                    result.get("isError").and_then(Value::as_bool).unwrap_or(false);
                let output = if text.is_empty() {
                    serde_json::to_string(&result).unwrap_or_default()
                } else {
                    text
                };
                (output, is_error)
            }
            Err(e) => (e.to_string(), true),
        }
    }

    async fn stop(&self) {
        if let Some(mut child) = self.child.lock().await.take() {
            child.kill().await.ok();
            child.wait().await.ok();
        }
    }
}

// ── Public plane ──────────────────────────────────────────────────────────────

/// `ExecutionPlane` that proxies tool calls to MCP servers over JSON-RPC 2.0 (stdio).
///
/// Credentials live in a `CredentialVault` and are injected into each server's subprocess
/// environment — the model never sees the credential values.
///
/// Usage:
/// ```ignore
/// let mut plane = McpProxyPlane::new(Arc::new(EnvCredentialVault));
/// plane.add_server("brave", McpServerConfig {
///     command: "npx".into(),
///     args: vec!["-y".into(), "@modelcontextprotocol/server-brave-search".into()],
///     credential_keys: vec!["BRAVE_API_KEY".into()],
///     env: Default::default(),
/// });
/// plane.connect().await?;
/// // ... use with RuntimeRunner ...
/// plane.disconnect().await;
/// ```
pub struct McpProxyPlane {
    server_configs: HashMap<String, McpServerConfig>,
    vault: Arc<dyn CredentialVault>,
    connections: Vec<Arc<McpConnection>>,
    tool_to_conn: HashMap<String, Arc<McpConnection>>,
    local: LocalExecutionPlane,
    local_names: HashSet<String>,
}

impl McpProxyPlane {
    pub fn new(vault: Arc<dyn CredentialVault>) -> Self {
        Self {
            server_configs: HashMap::new(),
            vault,
            connections: Vec::new(),
            tool_to_conn: HashMap::new(),
            local: LocalExecutionPlane::new(),
            local_names: HashSet::new(),
        }
    }

    pub fn add_server(&mut self, name: impl Into<String>, config: McpServerConfig) -> &mut Self {
        self.server_configs.insert(name.into(), config);
        self
    }

    /// Start all configured MCP server processes and discover their tool schemas.
    pub async fn connect(&mut self) -> Result<()> {
        for (name, config) in &self.server_configs {
            let conn = McpConnection::start(name, config, self.vault.as_ref()).await?;
            for schema in &conn.schemas {
                self.tool_to_conn.insert(schema.name.to_string(), Arc::clone(&conn));
            }
            self.connections.push(conn);
        }
        Ok(())
    }

    /// Gracefully stop all MCP server processes.
    pub async fn disconnect(&mut self) {
        for conn in &self.connections {
            conn.stop().await;
        }
        self.connections.clear();
        self.tool_to_conn.clear();
    }

    /// Register an in-process tool. Local tools take priority over MCP tools of the same name.
    pub fn register(&mut self, tool: RegisteredTool) -> &mut Self {
        self.local_names.insert(tool.schema.name.to_string());
        self.local.register(tool);
        self
    }

    pub fn unregister(&mut self, name: &str) -> &mut Self {
        self.local_names.remove(name);
        self.local.unregister(name);
        self
    }
}

impl ExecutionPlane for McpProxyPlane {
    fn schemas(&self) -> Vec<ToolSchema> {
        let mut result = self.local.schemas();
        for conn in &self.connections {
            result.extend(conn.schemas.iter().cloned());
        }
        result
    }

    fn execute_all<'a>(
        &'a self,
        calls: &'a [ToolCall],
        ctx: RunContext<'a>,
    ) -> Pin<Box<dyn Stream<Item = Result<RunEvent>> + Send + 'a>> {
        Box::pin(try_stream! {
            let RunContext {
                agent_id, skill_dir, dream_store, knowledge_source, governance, on_tool_suspend,
            } = ctx;

            let local_calls: Vec<_> = calls
                .iter()
                .filter(|c| self.local_names.contains(c.name.as_str()))
                .cloned()
                .collect();
            let mcp_calls: Vec<_> = calls
                .iter()
                .filter(|c| !self.local_names.contains(c.name.as_str()))
                .cloned()
                .collect();

            if !local_calls.is_empty() {
                let local_ctx = RunContext {
                    agent_id,
                    skill_dir,
                    dream_store,
                    knowledge_source,
                    governance: governance.clone(),
                    on_tool_suspend: on_tool_suspend.clone(),
                };
                let mut s = self.local.execute_all(&local_calls, local_ctx);
                while let Some(evt) = s.next().await {
                    yield evt?;
                }
            }

            // Group MCP calls by connection; concurrent across servers, sequential within.
            let mut groups: HashMap<usize, (Arc<McpConnection>, Vec<ToolCall>)> = HashMap::new();
            let mut unknown: Vec<ToolCall> = Vec::new();
            for call in &mcp_calls {
                if let Some(conn) = self.tool_to_conn.get(call.name.as_str()) {
                    let key = Arc::as_ptr(conn) as usize;
                    let entry = groups.entry(key).or_insert_with(|| (Arc::clone(conn), Vec::new()));
                    entry.1.push(call.clone());
                } else {
                    unknown.push(call.clone());
                }
            }

            for call in &unknown {
                yield RunEvent::ToolResult {
                    call_id: call.id.to_string(),
                    content: format!("unknown MCP tool: {}", call.name),
                    is_error: true,
                };
            }

            let futs: Vec<_> = groups
                .into_values()
                .map(|(conn, group_calls)| async move {
                    let mut results = Vec::new();
                    for call in &group_calls {
                        let (output, is_error) = conn.execute(call).await;
                        results.push((call.id.to_string(), output, is_error));
                    }
                    results
                })
                .collect();

            let all_results = futures::future::join_all(futs).await;
            for group in all_results {
                for (call_id, content, is_error) in group {
                    yield RunEvent::ToolResult { call_id, content, is_error };
                }
            }
        })
    }
}

File: node/src/runtime/mcp-proxy-plane.ts
import { spawn } from "node:child_process"
import { createInterface } from "node:readline"
import type { ToolCall, ToolSchema, StreamEvent, ContentBlock } from "../types.js"
import type { ToolResultEvent } from "../types.js"
import type { RegisteredTool } from "../tools/index.js"
import type { ExecutionPlane, RunContext } from "./execution-plane.js"
import { LocalExecutionPlane } from "./execution-plane.js"
import type { CredentialVault } from "./credential-vault.js"
import { formatToolError } from "../tools/errors.js"
import { operationAbortSignal } from "./reliability.js"

/** Raw MCP `tools/call` response `content` block shapes we accept (text/image/audio). */
export interface McpContentBlock {
  type: string
  text?: string
  data?: string
  mimeType?: string
}

/**
 * spc_012-N-02: pure conversion, unit-testable without spawning a subprocess. Prior behavior
 * silently dropped every non-`"text"` block (INV-012-01 violation — an MCP tool returning a
 * screenshot lost the image with no trace). `output` stays the text projection for backward
 * compatibility with the 24 existing call sites — when non-text blocks exist the projection
 * carries an explicit `[modality]` placeholder, so text-only providers (spc_012-N-04) degrade visibly instead
 * of swallowing the block. `contentParts` is populated (and only populated) when the response
 * actually contains a non-text block, so pure-text tool calls see zero behavior change.
 */
export function mcpResultToToolOutput(result: {
  content?: McpContentBlock[]
  isError?: boolean
}): { output: string; isError: boolean; contentParts?: ContentBlock[] } {
  const blocks = result.content ?? []
  const hasNonText = blocks.some(c => c.type !== "text")
  if (!hasNonText) {
    const text = blocks
      .filter(c => c.type === "text")
      .map(c => c.text ?? "")
      .join("\n")
    return { output: text || JSON.stringify(result), isError: result.isError ?? false }
  }
  const output = blocks
    .map(c => (c.type === "text" ? c.text ?? "" : `[${c.type}]`))
    .join("\n")
  const contentParts: ContentBlock[] = blocks.map(c => {
    if (c.type === "text") return { type: "text", text: c.text ?? "" }
    if (c.type === "image") {
      return { type: "image", source: { kind: "base64", data: c.data ?? "" }, mediaType: c.mimeType }
    }
    if (c.type === "audio") {
      return { type: "audio", source: { kind: "base64", data: c.data ?? "" }, mediaType: c.mimeType ?? "audio/wav" }
    }
    return { type: "text", text: JSON.stringify(c) }
  })
  return { output, isError: result.isError ?? false, contentParts }
}

export interface McpServerConfig {
  /** Executable to run (e.g. "npx", "python3", "/usr/local/bin/my-mcp-server"). */
  command: string
  /** Command-line arguments. */
  args?: string[]
  /**
   * Keys to look up in the vault and inject as env vars with the same name.
   * The credential value is never exposed to the model — only to the subprocess.
   */
  credentialKeys?: string[]
  /** Additional static env vars forwarded to the subprocess. */
  env?: Record<string, string>
}

// ── Internal JSON-RPC / MCP client ───────────────────────────────────────────

interface RpcRequest { jsonrpc: "2.0"; method: string; params?: unknown; id: number }
interface RpcResponse { jsonrpc: "2.0"; result?: unknown; error?: { code: number; message: string }; id: number }

class McpConnection {
  private child: ReturnType<typeof spawn> | null = null
  private pending = new Map<number, { resolve(r: unknown): void; reject(e: Error): void }>()
  private nextId = 1
  private _schemas: ToolSchema[] = []
  private schemaNames = new Set<string>()

  constructor(
    readonly serverName: string,
    private readonly config: McpServerConfig,
    private readonly vault: CredentialVault,
  ) {}

  async start(): Promise<void> {
    const env: Record<string, string> = { ...(process.env as Record<string, string>), ...(this.config.env ?? {}) }
    for (const key of this.config.credentialKeys ?? []) {
      const val = await this.vault.get(key)
      if (val !== undefined) env[key] = val
    }

    this.child = spawn(this.config.command, this.config.args ?? [], {
      env,
      stdio: ["pipe", "pipe", "inherit"],
    })

    const rl = createInterface({ input: this.child.stdout!, crlfDelay: Infinity })
    rl.on("line", line => {
      if (!line.trim()) return
      try {
        const msg = JSON.parse(line) as RpcResponse
        const cb = this.pending.get(msg.id)
        if (!cb) return
        this.pending.delete(msg.id)
        if (msg.error) cb.reject(new Error(`MCP(${this.serverName}) ${msg.error.code}: ${msg.error.message}`))
        else cb.resolve(msg.result)
      } catch { /* ignore malformed lines */ }
    })

    // MCP handshake
    await this.request("initialize", {
      protocolVersion: "2024-11-05",
      capabilities: { tools: {} },
      clientInfo: { name: "deepstrike", version: "0.1.0" },
    })
    // initialized is a notification (no id, no response expected)
    this.notify("notifications/initialized")

    const listResult = await this.request("tools/list") as {
      tools?: Array<{ name: string; description?: string; inputSchema?: Record<string, unknown> }>
    }
    for (const t of listResult.tools ?? []) {
      const schema: ToolSchema = {
        name: t.name,
        description: t.description ?? t.name,
        parameters: JSON.stringify(t.inputSchema ?? { type: "object", properties: {} }),
      }
      this._schemas.push(schema)
      this.schemaNames.add(t.name)
    }
  }

  private request(method: string, params?: unknown, signal?: AbortSignal): Promise<unknown> {
    return new Promise((resolve, reject) => {
      if (!this.child?.stdin) { reject(new Error(`MCP server "${this.serverName}" not running`)); return }
      const id = this.nextId++
      const cleanup = () => signal?.removeEventListener("abort", abort)
      const abort = () => {
        this.pending.delete(id)
        cleanup()
        reject(signal?.reason instanceof Error ? signal.reason : new Error("operation cancelled"))
      }
      this.pending.set(id, {
        resolve: result => { cleanup(); resolve(result) },
        reject: error => { cleanup(); reject(error) },
      })
      signal?.addEventListener("abort", abort, { once: true })
      if (signal?.aborted) { abort(); return }
      const msg: RpcRequest = { jsonrpc: "2.0", method, id }
      if (params !== undefined) msg.params = params
      this.child.stdin.write(JSON.stringify(msg) + "\n")
    })
  }

  private notify(method: string, params?: unknown): void {
    if (!this.child?.stdin) return
    const msg: Record<string, unknown> = { jsonrpc: "2.0", method }
    if (params !== undefined) msg.params = params
    this.child.stdin.write(JSON.stringify(msg) + "\n")
  }

  schemas(): ToolSchema[] { return this._schemas }
  hasSchema(name: string): boolean { return this.schemaNames.has(name) }

  async execute(
    call: ToolCall,
    signal?: AbortSignal,
  ): Promise<{ output: string; isError: boolean; contentParts?: ContentBlock[] }> {
    try {
      const args = JSON.parse(call.arguments || "{}") as Record<string, unknown>
      const result = await this.request("tools/call", { name: call.name, arguments: args }, signal) as {
        content?: McpContentBlock[]
        isError?: boolean
      }
      return mcpResultToToolOutput(result)
    } catch (err) {
      return { output: formatToolError(err), isError: true }
    }
  }

  async stop(): Promise<void> {
    this.child?.kill()
    this.child = null
    for (const cb of this.pending.values()) cb.reject(new Error(`MCP server "${this.serverName}" stopped`))
    this.pending.clear()
  }
}

// ── Public plane ──────────────────────────────────────────────────────────────

/**
 * ExecutionPlane that proxies tool calls to MCP servers.
 *
 * Credentials live in a CredentialVault and are injected into each server's
 * subprocess environment — the model never sees the credential values.
 *
 * Usage:
 *   const plane = new McpProxyPlane({
 *     servers: { brave: { command: "npx", args: ["-y", "@modelcontextprotocol/server-brave-search"], credentialKeys: ["BRAVE_API_KEY"] } },
 *     vault: new EnvCredentialVault(),
 *   })
 *   await plane.connect()
 *   // ... use with RuntimeRunner ...
 *   await plane.disconnect()
 */
export class McpProxyPlane implements ExecutionPlane {
  private connections = new Map<string, McpConnection>()
  private toolToConn = new Map<string, McpConnection>()
  private localPlane = new LocalExecutionPlane()
  private localNames = new Set<string>()

  constructor(private readonly opts: {
    servers: Record<string, McpServerConfig>
    vault: CredentialVault
    timeoutMs?: number
  }) {}

  /** Start all configured MCP server processes and discover their tool schemas. */
  async connect(): Promise<void> {
    for (const [name, config] of Object.entries(this.opts.servers)) {
      const conn = new McpConnection(name, config, this.opts.vault)
      await conn.start()
      this.connections.set(name, conn)
      for (const schema of conn.schemas()) this.toolToConn.set(schema.name, conn)
    }
  }

  /** Gracefully stop all MCP server processes. */
  async disconnect(): Promise<void> {
    for (const conn of this.connections.values()) await conn.stop()
    this.connections.clear()
    this.toolToConn.clear()
  }

  register(...tools: RegisteredTool[]): this {
    this.localPlane.register(...tools)
    for (const t of tools) this.localNames.add(t.schema.name)
    return this
  }

  unregister(name: string): this {
    this.localPlane.unregister(name)
    this.localNames.delete(name)
    return this
  }

  schemas(): ToolSchema[] {
    const mcp: ToolSchema[] = []
    for (const conn of this.connections.values()) mcp.push(...conn.schemas())
    return [...this.localPlane.schemas(), ...mcp]
  }

  async *executeAll(calls: ToolCall[], ctx: RunContext): AsyncIterable<StreamEvent> {
    const localCalls = calls.filter(c => this.localNames.has(c.name))
    const mcpCalls   = calls.filter(c => !this.localNames.has(c.name))

    if (localCalls.length > 0) yield* this.localPlane.executeAll(localCalls, ctx)

    // MCP calls are inherently sequential per server (JSON-RPC request/response)
    // but different servers can be called concurrently.
    const groups = new Map<McpConnection, ToolCall[]>()
    const unknown: ToolCall[] = []
    for (const call of mcpCalls) {
      const conn = this.toolToConn.get(call.name)
      if (!conn) { unknown.push(call); continue }
      if (!groups.has(conn)) groups.set(conn, [])
      groups.get(conn)!.push(call)
    }

    for (const call of unknown) {
      yield { type: "tool_result", callId: call.id, name: call.name, content: `unknown MCP tool: ${call.name}`, isError: true } as ToolResultEvent
    }

    const tasks = Array.from(groups.entries()).map(async ([conn, serverCalls]) => {
      const results: Array<{
        call: ToolCall
        result: { output: string; isError: boolean; contentParts?: ContentBlock[] }
      }> = []
      for (const call of serverCalls) {
        const signal = operationAbortSignal(ctx.operation, this.opts.timeoutMs ?? 30_000)
        results.push({ call, result: await conn.execute(call, signal) })
      }
      return results
    })

    for (const settled of await Promise.all(tasks)) {
      for (const { call, result } of settled) {
        yield {
          type: "tool_result",
          callId: call.id,
          name: call.name,
          content: result.output,
          isError: result.isError,
          contentParts: result.contentParts,
        } as ToolResultEvent
      }
    }
  }
}
